modules/RTC/RTCUtils.js

```diff
@@ -1287,167 +1287,147 @@ class RTCUtils extends Listenable {
+        // support MediaStream constructors as defined by
+        // https://www.w3.org/TR/mediacapture-streams/#constructors. Instead it
+        // has a single constructor which expects an NSNumber as a MediaStream
+        // ID. Temasys does not support the constructor at all. So we get audio
+        // and video in two distinct GUM calls and pass them back into a single
+        // object.
+        if (browser.isReactNative() || browser.isTemasysPluginUsed()) {
+            gumPromise = new Promise((resolve, reject) => {
                 const deviceGUM = {
-                    'audio': GUM.bind(self, [ 'audio' ]),
-                    'video': GUM.bind(self, [ 'video' ])
+                    audio: (...args) =>
+                        this.getUserMediaWithConstraints([ 'audio' ], ...args),
+                    video: (...args) =>
+                        this.getUserMediaWithConstraints([ 'video' ], ...args),
+                    desktop: (...args) =>
+                        screenObtainer.obtainStream(dsOptions, ...args)
                 };
 
-                if (screenObtainer.isSupported()) {
-                    deviceGUM.desktop = screenObtainer.obtainStream.bind(
-                        screenObtainer,
-                        dsOptions);
-                }
-
-                // With FF/IE we can't split the stream into audio and video
-                // because FF doesn't support media stream constructors. So, we
-                // need to get the audio stream separately from the video stream
-                // using two distinct GUM calls. Not very user friendly :-( but
-                // we don't have many other options neither.
-                //
-                // Note that we pack those 2 streams in a single object and pass
-                // it to the successCallback method.
                 obtainDevices({
                     devices: options.devices,
                     streams: [],
-                    successCallback,
+                    successCallback: resolve,
                     errorCallback: reject,
                     deviceGUM
                 });
-            } else {
-                const hasDesktop = options.devices.indexOf('desktop') > -1;
+            });
+        } else {
+            options.devices = options.devices.filter(device =>
+                device !== 'desktop');
+
+            const promiseWrappedGum = function(requestedDevices, gumOptions) {
+                return new Promise((resolve, reject) => {
+
+                    if (requestedDevices.length) {
+                        this.getUserMediaWithConstraints(
+                            requestedDevices,
+                            resolve,
+                            reject,
+                            gumOptions);
+                    } else {
+                        resolve(null);
+                    }
+                });
+            }.bind(this);
```

Comment:
will arrow function work here to not do .bind(this) ? it used to be preferred way